src/mysky/encrypted_files.ts

```diff
@@ -0,0 +1,330 @@
+ *
+ *  88 KiB     96 KiB    104 KiB    112 KiB    120 KiB
+ * 128 KiB    136 KiB    144 KiB    152 KiB    160 KiB
+ *
+ * 176 KiB    192 Kib    208 KiB    224 KiB    240 KiB
+ * 256 KiB    272 KiB    288 KiB    304 KiB    320 KiB
+ *
+ * 352 KiB    ... etc
+ * ```
+ *
+ * Note that the first 20 valid sizes are all a multiple of 4 KiB, the next 10
+ * are a multiple of 8 KiB, and each 10 after that the multiple doubles. We use
+ * this method of padding files to prevent an adversary from guessing the
+ * contents or structure of the file based on its size.
+ *
+ * @param initialSize - The size of the file.
+ * @returns - The final size, padded to a pad block.
+ * @throws - Will throw if the size would overflow the JS number type.
+ */
+export function padFileSize(initialSize: number): number {
+  const kib = 1 << 10;
+  for (let n = 0; ; n++) {
+    // Prevent overflow. Max JS number size is 2^53-1.
+    if (n >= 53) {
+      throw new Error("Could not pad file size, overflow detected.");
+    }
+    if (initialSize <= (1 << n) * 80 * kib) {
+      const paddingBlock = (1 << n) * 4 * kib;
+      let finalSize = initialSize;
+      if (finalSize % paddingBlock != 0) {
+        finalSize = initialSize - (initialSize % paddingBlock) + paddingBlock;
+      }
+      return finalSize;
+    }
+  }
+}
+
+/**
+ * Checks if the given size corresponds to the correct padded block.
+ *
+ * @param size - The given file size.
+ * @returns - Whether the size corresponds to a padded block.
+ * @throws - Will throw if the size would overflow the JS number type.
+ */
+export function checkPaddedBlock(size: number): boolean {
+  const kib = 1 << 10;
+  for (let n = 0; ; n++) {
+    // Prevent overflow. Max JS number size is 2^53-1.
+    if (n >= 53) {
+      throw new Error("Could not check padded file size, overflow detected.");
+    }
+    if (size <= (1 << n) * 80 * kib) {
+      const paddingBlock = (1 << n) * 4 * kib;
+      return size % paddingBlock === 0;
+    }
+  }
+}
+
+/**
+ * Decodes the encoded encrypted file metadata.
+ *
+ * @param bytes - The encoded metadata.
+ * @returns - The decoded metadata.
+ * @throws - Will throw if the given bytes are of the wrong length.
+ */
+function decodeEncryptedFileMetadata(bytes: Uint8Array): EncryptedFileMetadata {
+  // Ensure input is correct length.
+  validateUint8ArrayLen("bytes", bytes, "encrypted file metadata bytes", ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
+
+  const version = bytes[0];
+
+  return {
+    version,
+  };
+}
+
+/**
+ * Encodes the given encrypted file metadata.
+ *
+ * @param metadata - The given metadata.
+ * @returns - The encoded metadata bytes.
+ * @throws - Will throw if the version does not fit in a byte.
+ */
+export function encodeEncryptedFileMetadata(metadata: EncryptedFileMetadata): Uint8Array {
+  const bytes = new Uint8Array(ENCRYPTION_HIDDEN_FIELD_METADATA_LENGTH);
+
+  // Encode the version
+  if (metadata.version >= 1 << 8) {
```

Comment:
I know `1 << 8` equals `256` but let's not overcomplicate simple code with byte shifts

```suggestion
  if (metadata.version > 255) {
```

Reply:
I thought `1 << 8` would emphasize that the version should fit in an 8-bit type. Anyway, not fixing due to extremely low severity.

Reply:
I agree with Karol here - byte shifts are fine when you need them but most people (myself included) always make a pause when they encounter them, be it just to make sure that they read them correctly.

This is one of those cases where the code is very clever and that hurts readability. I wouldn't block a PR for this but I wouldn't call this low severity, either.

Reply:
Yeah I will have to insist here @m-cat - byte shifts should be avoided unless necessary, I know they might be more common in go, c++ or java but they are very unusual thing to use in javascript.

There is even an eslint rule that suggest disabling them completely unless you explicitly need them on one line and you enable them there specifically. https://eslint.org/docs/rules/no-bitwise

Reply:
Fair enough, I wrote up https://github.com/SkynetLabs/skynet-js/issues/187. However this is not a blocking issue for this PR.